content/blog/2025-03-20-new-hash-table.md

```diff
@@ -0,0 +1,202 @@
++++
+title= "A new hash table"
+date= 2025-03-20 00:00:00
+description= "Designing a state-of-the art hash table implementation"
+authors= [ "zuiderkwast", "SoftlyRaining"]
++++
+
+Valkey is essentially a giant hash table attached to the network. A hash table
+is the data structure that maps keys to values. When optimizing for latency, CPU
+and memory usage, it's natural to look at the hash table internals. By replacing
+the hash table with a different implementation, we have managed to reduce the
+memory usage by roughly 20 bytes per key-value pair and improve the latency and
+CPU usage by rougly 10% for instances without I/O threading.
+
+Results
+-------
+
+Memory usage for keys of length N and value of length M bytes. TBD.
+
+| Version    | Memory usage per key   |
+|------------|------------------------|
+| Valkey 7.2 | ? bytes                |
+| Valkey 8.0 | ? bytes                |
+| Valkey 8.1 | ? bytes                |
+
+The benchmarks below were run using a key size of N and a value size of M bytes, without pipelining.
+
+| Command                 | Valkey 7.2 | Valkey 8.0 | Valkey 8.1 |
+|-------------------------|------------|------------|------------|
+| SET                     | Xµs, Y QPS | ?          | ?          |
+| GET                     | Xµs, Y QPS | ?          | ?          |
+| ...                     | ...        | ?          | ?          |
+
+The benchmark was run on an xxxx using yyyy, without I/O threads.
+
+Background
+----------
+
+The slowest operation when looking up a key-value pair is by far reading from
+the main RAM memory. A key point when optimizing a hash table is therefore to
+make sure we have as few memory accesses as possible. Ideally, the memory
+reading is already in the CPU cache, which is much faster memory that belong to
+the CPU.
+
+When optimizing for memory usage, we also want to minimize the number of
+allocation and pointers between them, because a pointer is 8 bytes in a 64-bit
+system. If we save one pointer per key-value pair, for 100 million keys that's
+almost a gigabyte.
+
+When a computer loads some data from the main memory into the CPU cache, it does
+so in blocks of one cache line. The cache-line size is 64 bytes on almost all
+modern hardware. Recent work on hash tables, such as "Swiss tables", are highly
+optimized for cache lines. When looking up a key, if it's not found where you
+first look for it (due to a hash collission), then it should ideally be found
```

Comment:
```suggestion
first look for it (due to a hash collision), then it should ideally be found
```
First introduction of concepts such as hash collisions. I mentioned earlier it would be helpful to give an overview of the current design, probably can walk through these concepts.

Reply:
Yeah, it's hard to know how much of computer science lecture we have to give. We'll have to assume some basic knowledge of what a hash table is, right?